lib/racecar/consumer_set.rb

```diff
@@ -18,24 +18,25 @@ def subscribe
     end
 
     def poll(timeout_ms)
-      current.poll(timeout_ms)
+      current.poll(timeout_ms).tap do |msg|
+        @consumer_iterator.next if msg.nil?
+      end
     rescue Rdkafka::RdkafkaError => e
       raise unless e.is_partition_eof?
-      @logger.debug "No more messages on this partition."
+      @logger.debug "No more messages on these partition(s)."
       @consumer_iterator.next
       nil
     end
 
     def batch_poll(timeout_ms)
       @batch_started_at = Time.now
       @messages = []
-      @messages << current.poll(timeout_ms) while collect_messages_for_batch?
-      @messages.compact
-    rescue Rdkafka::RdkafkaError => e
-      raise unless e.is_partition_eof?
-      @logger.debug "No more messages on this partition."
-      @consumer_iterator.next
-      @messages.compact
+      while collect_messages_for_batch? do
+        msg = poll(timeout_ms)
+        break if msg.nil?
+        @messages << msg
+      end
```

Comment:
Is this something we can push into rdkafka or librdkafka?

Reply:
I'm not sure I understood exactly what you're referring to, but I have the following suggestion:

librdkafka recently switched the default for `enable.partition.eof` to `false` https://github.com/edenhill/librdkafka/commit/a869fa19359fc285b19496c4321abe199322a736#diff-ab78e771fd78692a15ed791e0269a2db

This means there will be no more partition eof errors whatsoever and racecar would just have to deal with the `nil` case. For the transition period racecar should simply disable `enable.partition.eof` itself. Does that sound reasonable?

That being said, there's https://github.com/appsignal/rdkafka-ruby/issues/12 about the same issue. Considering it ought to be a simple wrapper around the C api of librdkafka, I'm not sure adding a high level `each`/`poll` makes sense there.